src/app/server/Server.h

```diff
@@ -59,11 +62,80 @@ using ServerTransportMgr = chip::TransportMgr<chip::Transport::UDP
 #endif
                                               >;
 
+struct ServerInitParams
+{
+    ServerInitParams()          = default;
+    virtual ~ServerInitParams() = default;
+
+    // Not copyable
+    ServerInitParams(const ServerInitParams &) = delete;
+    ServerInitParams & operator=(const ServerInitParams &) = delete;
+
+    // App delegate to handle some lifecycle events
+    AppDelegate * appDelegate = nullptr;
+    // Port to use for Matter commissioning/operational traffic
+    uint16_t operationalServicePort = CHIP_PORT;
+    // Port to use for UDC if supported
+    uint16_t userDirectedCommissioningPort = CHIP_UDC_PORT;
+    // Interface on which to run daemon
+    Inet::InterfaceId interfaceId = Inet::InterfaceId::Null();
+
+    // Persistent storage delegate: MUST be injected. Used to maintain storage by much common code.
+    PersistentStorageDelegate * persistentStorageDelegate = nullptr;
+    // Group data provider: MUST be injected. Used to maintain critical keys such as the Identity
+    // Protection Key (IPK) for case
+    Credentials::GroupDataProvider * groupDataProvider = nullptr;
+    // Access control delegate: MUST be injected. Used to look-up access control rules
+    Access::AccessControl::Delegate * accessDelegate = nullptr;
+};
+
+// Transitional pre-injection version of ServierInitParams
+struct CommonCaseDeviceServerInitParams : public ServerInitParams
+{
+    CommonCaseDeviceServerInitParams() = default;
+
+    // Not copyable
+    CommonCaseDeviceServerInitParams(const CommonCaseDeviceServerInitParams &) = delete;
+    CommonCaseDeviceServerInitParams & operator=(const CommonCaseDeviceServerInitParams &) = delete;
+
+    virtual CHIP_ERROR InitBeforeServerInit()
+    {
+        static chip::KvsPersistentStorageDelegate sKvsPersistenStorageDelegate;
+        static chip::Credentials::GroupDataProviderImpl sGroupDataProvider;
+
+        // KVS-based persistent storage delegate injection
+        chip::DeviceLayer::PersistedStorage::KeyValueStoreManager & kvsManager = DeviceLayer::PersistedStorage::KeyValueStoreMgr();
+        ReturnErrorOnFailure(sKvsPersistenStorageDelegate.Init(&kvsManager));
+        this->persistentStorageDelegate = &sKvsPersistenStorageDelegate;
+
+        // Group Data provider injection
+        sGroupDataProvider.SetStorageDelegate(&sKvsPersistenStorageDelegate);
+        ReturnErrorOnFailure(sGroupDataProvider.Init());
+        this->groupDataProvider = &sGroupDataProvider;
+
+        // Inject access control delegate
+        this->accessDelegate = Access::Examples::GetAccessControlDelegate(&sKvsPersistenStorageDelegate);
+
+        return CHIP_NO_ERROR;
+    }
+
+    virtual void Shutdown()
```

Comment:
When would one call this?  And if we expect people to call this, shouldn't we be transferring ownership of those things to Server if it will _also_ call these Finish functions?

Reply:
Commented. Created issue #16968